### PR TITLE
New contract interpreter based on the WAMR WASM interpreter

### PR DESCRIPTION
**Pull request by cmickeyb**
This PR introduces an experimental WASM-based contract interpreter called Wawaka. The interpreter uses the Web Assembly Micro Runtime (https://github.com/bytecodealliance/wasm-micro-runtime). In addition to the interpreter, several sample contracts are provided (written in C++) and a small library of useful functions.

### CODE REVIEW

File: common/interpreter/wawaka_wasm/README.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+cd ${PDO_SOURCE_ROOT}/contracts/wawaka
+pdo-test-contract --no-ledger --interpreter wawaka --contract mock-contract \
+    --expressions ./mock-contract/test-long.exp --loglevel info
+```
+
+## Basics of a Contract ##
+
+Note that compilation into WASM that will run in the contract enclave can be somewhat tricky. Specifically, all symbols whether used or not must be bound. The wawaka interpreter will fail if it attempts to load WASM code with unbound symbols.
+
+It may be easiest to copy the
```
Comment (prakashngit) [inline common/interpreter/wawaka_wasm/README.md]: You may want to just say here ("User documentation will follow soon").

Does it make sense to attach your slides on WAWAKA,  I find them helpful for an overview of the process.

**Comment by cmickeyb** [comment]
> I was able to run `pdo-test-contract` but I think I have encountered few minor problems:
> 
> * PDO_SOURCE_ROOT is set but not exported, so compilation failed
> * there is an assumption that PDO is already compiled and you want to rebuild it for wasm -- by just rebiulding from a clean install it could not find the configuration files

please provide more information. i build without PDO_SOURCE_ROOT set (that is, i'm explicitly unsetting it before i source common-config.sh and then unset it again before running build). and i always build cleanly. no issues. are you sure you're sourcing common-config.sh?

File: common/interpreter/wawaka_wasm/README.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+There are many toolchains that could be used to build a WASM code. We have tested (and our sample
+and test contracts use) [emscripten](https://emscripten.org/). Note that we currently require the `fastcomp` compiler which is no longer the default compiler.
+
+```bash
+cd ${PDO_SOURCE_ROOT}
+
+git clone https://github.com/emscripten-core/emsdk.git
+cd ${PDO_SOURCE_ROOT}/emsdk
+
```
[the remark this answers is not in the document]
Comment (cmickeyb) [reply common/interpreter/wawaka_wasm/README.md]: you can put it anywhere you want. but this is a clear set of instructions to get going.

File: common/interpreter/wawaka_wasm/README.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+### Clone the WAMR source ###
+
+If wawaka is configured as the contract interpreter, the libraries implementing the WASM interpreter
+will be built for use with Intel SGX. The source for the WAMR interpreter is distributed separately and must be downloaded.
+
+```bash
+cd ${PDO_SOURCE_ROOT}
+git clone https://github.com/intel/wasm-micro-runtime.git wasm
+```
```
[the remark this answers is not in the document]
Comment (cmickeyb) [reply common/interpreter/wawaka_wasm/README.md]: same as for emsdk

File: eservice/pdo/eservice/pdo_enclave.py [hunk 4]
```diff
@@ -277,6 +277,7 @@ def create_signup_info(originator_public_key_hash, nonce):
     #
     # Start building up the signup info dictionary we will serialize
     signup_info = {
+        'interpreter' : signup_data['interpreter'],
         'verifying_key': signup_data['verifying_key'],
         'encryption_key': signup_data['encryption_key'],
         'proof_data': 'Not present',
```
[the remark this answers is not in the document]
Comment (cmickeyb) [reply eservice/pdo/eservice/pdo_enclave.py]: at this point there is one MRENCLAVE so all enclaves will have the same interpreter. we can generalize later (i.e. having some enclaves registered as wawaka and some as gipsy)

File: common/interpreter/wawaka_wasm/README.md (hunk 1 above)
Comment (cmickeyb) [reply common/interpreter/wawaka_wasm/README.md]: Clearly we have to mature this. I don't think the slides are necessary just yet.

File: common/interpreter/wawaka_wasm/README.md [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+### Test the Configuration ###
+
+Sample wawaka contracts are built and installed along with the
+interpreter. You can run the simple test contract as follows:
+
+```bash
+cd ${PDO_SOURCE_ROOT}/contracts/wawaka
+pdo-test-contract --no-ledger --interpreter wawaka --contract mock-contract \
+    --expressions ./mock-contract/test-long.exp --loglevel info
+```
```
[the remark this answers is not in the document]
Comment (cmickeyb) [reply common/interpreter/wawaka_wasm/README.md]: Pulling the default interpreter from PDO_INTERPRETER

**Pull request merged**